feat: copy command

mmalenic opened the pull request:
Closes #36

### Changes
* Adds basic copy command for single-part copies only, from S3->S3, local->S3, S3->local and local->local.
    * Two modes are available, server-side copy and download-upload.
    * Tags/metadata can be copied, suppressed, or best-effort copied (ignored access denied errors).

mmalenic commented on cloud-checksum/src/io/copy/aws.rs:
```diff
@@ -0,0 +1,244 @@
+        }
+    }
+
+    fn is_access_denied<T: ProvideErrorMetadata>(err: &SdkError<T, HttpResponse>) -> bool {
+        if let Some(err) = err.as_service_error() {
+            if err
+                .code()
+                .is_some_and(|code| code == "AccessDenied" || code == "InvalidSecurity")
+            {
+                return true;
+            }
+        }
+
+        false
+    }
```
I'm not sure that this is the best way to handle this feature. There's three options - copy tags and fail, try copy tags but don't fail if permission is denied, and don't copy tags.

This will also ignore other permission errors, such as `PutObject` or `GetObject`, which might be unintuitive if the CLI succeeds but the copy failed. I think it might be better to just remove the "best-effort" option?

Otherwise, I think this CLI should do the copy, and then actually perform a "check" operation on the output to confirm that it is valid. 

I'm not sure if we want this in general or if it's better to leave this as a separate command. E.g. do we want to do `cloud-checksum copy <src> <dest>` or `cloud-checksum copy <src> <dest> && cloud-checksum check <src> <dest>` to confirm a valid copy?

brainstorm replied:
I'd keep it as deterministic as possible: copy tags vs no copy tags and report errors if either fails (in any object)... the best effort way is an unnecessary rabbit whole I reckon? :-S

andrewpatto replied:
I think we have to do multi part first before deciding. Given some of the tag setting is going to be "extra" code - lets see how that plays out catering for different tag scenarios.

Maybe roman is right that the basic copy should just be deterministic - but there could be a flag ("--explicit-destination-check") that adds in a check after the copy

andrewpatto commented on cloud-checksum/src/lib.rs:
```diff
@@ -60,74 +57,157 @@ impl Commands {
+    /// how the source was uploaded. This can be used to override that.
+    #[arg(short, long, env, value_parser = |s: &str| parse_size(s))]
+    pub part_size: Option<u64>,
+}
```
Not at all needed for this PR - but I think the eventual mode we will prefer/have a flag for is where threshold and part size are determined by reference to the details we can derive from the source object itself.

mmalenic replied:
Yes that's the intent. I was thinking that this flag would only be used to override any of that automatic detection?

andrewpatto commented:
Should tags/metadata be conflated as one thing - or should there be separate flags for controlling the behaviour of each? (not for this PR - but one to consider).

For instance for UMCCR copy out - we might very well want attributes copied but not tags. Also - we may want selective copying of _some_ attributes or tags. I will make a separate feature issue for this.

mmalenic commented:
> or should there be separate flags for controlling the behaviour of each? (not for this PR - but one to consider).

I thought about that - easy enough to add another flag to separate metadata and tags. I wasn't sure how well that distinction would translate to other cloud providers. I think Google cloud storage has a similar thing?

andrewpatto commented:
> I thought about that - easy enough to add another flag to separate metadata and tags. I wasn't sure how well that distinction would translate to other cloud providers. I think Google cloud storage has a similar thing?

I think this is our strength though - we _should_ think through what is means and how it all works - for a limited subset of clouds that we use. And then give the users those fine-grained options that `rclone` is never going to support.

So possibly we might need a field like "--translate-s3-tag-to-attribute" or something - that handles this situation for S3 compatible stores without tagging. Certainly not a high priority - but that's the vibe I am thinking.

mmalenic commented:
### Update
* Added options to copy tags and metadata separately in the CLI.
* Implemented retry logic for AccessDenied errors due to failed `PutObjectTagging` permissions so that "copy" and "best-effort" modes work correctly
